### ISSUE TITLE
Create events model

### ISSUE THREAD

**Issue by khvalov** [centrifugal/centrifuge-js#89]
It will be nice to add event model for managing token expiration like: 

centrifuge.on('tokenExpired', function(e){
....//some logic 
centrifuge.setToken('....');
...//some logic
 });

**Comment by FZambia** [comment]
Hi, please describe why current refresh behavior does not work for you and how suggested functionality will solve a problem.

**Comment by khvalov** [comment]
Actually I have to have SSO located on other than centrifuge domain. 
The concerns I have
- CORS (it can be solved using headers) 
- redirects. It can't be solved (at least in Chrome). If user session TTL expired I would like to popup a login credentials window rather have just session expired exception. 
Currently error I have: 
Access to XMLHttpRequest at 'http://api.zzz.local/token/refresh?uuid=3efe3c40-c0fa-11e8-b3f4-1c51435d0814' from origin 'http://cp.flock.local' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: **Redirect is not allowed for a preflight request**.

**Comment by khvalov** [comment]
As possible alternative you have to implement setting like 
`xhrFields: { withCredentials: true }, ` (http://promincproductions.com/blog/cross-domain-ajax-request-cookies-cors/)  to make sure that CORS request handled properly for auth purposes. 
Same approach I guess should work for joining private channels. Otherwise there is no way to auth user using SSO located out of the centrifuge located domain.

**Comment by FZambia** [comment]
I think you just need to properly configure CORS on your refresh endpoint. It's hard to say more about error you get without diving deep into your setup and trying to reproduce - but looks like there is some redirect involved which is not allowed.

Also see `onRefresh` callback https://github.com/centrifugal/centrifuge-js#refreshendpoint which allows to use custom code to get token - you can use whatever you want with it This is pretty close to your original proposal i suppose.

**Comment by khvalov** [comment]
Actualy I have solved my problem with CORS. But seems if nice feature to have. There a lot of SSO solutions have time-to-live for tokens. If it's just get expired - nice to have any mechanism to show login screen to user. Am I right ?

**Comment by khvalov** [comment]
onRefresh looks like exactly what I looking for. Thanks for advise

**Comment by FZambia** [comment]
Closing - if you have related questions feel free to ask here.
